src/libopensc/card-piv.c

```diff
@@ -507,209 +843,1967 @@ static struct sc_card_driver piv_drv = {
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* if intermediate sm_in_cvc present, cert signed it and sm_cvc is signed by sm_in_cvc */
+	if (priv->sm_flags & PIV_SM_FLAGS_SM_IN_CVC_PRESENT) {
+		r = piv_sm_verify_sig(card, cs->kdf_md(), cert_pkey,
+				priv->sm_in_cvc.body, priv->sm_in_cvc.bodylen,
+				priv->sm_in_cvc.signature,priv->sm_in_cvc.signaturelen);
+		if (r < 0) {
+			sc_log(card->ctx,"sm_in_cvc signature invalid");
+			r =  SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+		if ((in_cvc_group = EC_GROUP_new_by_curve_name(cs->nid)) == NULL
+				|| (in_cvc_pkey = EVP_PKEY_new()) == NULL
+				|| (in_cvc_eckey = EC_KEY_new_by_curve_name(cs->nid)) == NULL
+				|| (in_cvc_point = EC_POINT_new(in_cvc_group)) == NULL
+				|| EC_POINT_oct2point(in_cvc_group, in_cvc_point,
+					priv->sm_in_cvc.publicPoint, priv->sm_in_cvc.publicPointlen, NULL) <= 0
+				|| EC_KEY_set_public_key(in_cvc_eckey, in_cvc_point) <= 0
+				|| EVP_PKEY_set1_EC_KEY(in_cvc_pkey, in_cvc_eckey) != 1) {
+			sc_log(card->ctx, "OpenSSL failed to set EC pubkey, during verify");
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+#else
+		params_n = 0;
+		params[params_n++] = OSSL_PARAM_construct_utf8_string("group", cs->curve_group, 0);
+		params[params_n++] = OSSL_PARAM_construct_octet_string("pub",
+				priv->sm_in_cvc.publicPoint, priv->sm_in_cvc.publicPointlen);
+		params[params_n] = OSSL_PARAM_construct_end();
+
+		if (!(in_cvc_pkey_ctx = EVP_PKEY_CTX_new_from_name(PIV_LIBCTX, "EC", NULL))
+				|| !EVP_PKEY_fromdata_init(in_cvc_pkey_ctx)
+				|| !EVP_PKEY_fromdata(in_cvc_pkey_ctx, &in_cvc_pkey, EVP_PKEY_PUBLIC_KEY, params)
+				|| !in_cvc_pkey) {
+			sc_log(card->ctx, "OpenSSL failed to set EC pubkey, during verify");
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+#endif
+		r = piv_sm_verify_sig(card, cs->kdf_md(), in_cvc_pkey,
+				priv->sm_cvc.body, priv->sm_cvc.bodylen,
+				priv->sm_cvc.signature,priv->sm_cvc.signaturelen);
+
+	} else { /* cert signed  sm_cvc */
+		r = piv_sm_verify_sig(card, cs->kdf_md(), cert_pkey,
+				priv->sm_cvc.body, priv->sm_cvc.bodylen,
+				priv->sm_cvc.signature,priv->sm_cvc.signaturelen);
+	}
+	if (r < 0) {
+		sc_log(card->ctx,"sm_cvc signature invalid");
+		r =  SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* cert chain signatures match for oncard certs */
+	/* TODO check dates and other info as per 800-73-4 */
+
+	/* TODO check against off card CA chain if present,
+	 * Need opensc.conf options:
+	 *	where is CA cert chain?
+	 *	is it required?
+	 *	check for revocation?
+	 *	How often to check for revocation?
+	 *	When is SM used?
+	 *		Using NFC?
+	 *			(yes, main point of using SM)
+	 *		Should reading certificates be done in clear?
+	 *			(performance vs security)
+	 *		All crypto operations and PIN ?
+	 *			(yes for security)
+	 */
```

Comment:
Is it mendatory for some use case to check CVC signatures? If not, you may want to consider removing it.

I previously have implemented the same in sm-eac.c, because we can. But I later removed this, because this a) simplifies the source code, b) isn't useful if we don't have a trust anchor and c) implies managing a trust anchor by the user to make it actually meaningful (which nobody is interested in).

Reply:
> Is it mendatory for some use case to check CVC signatures? If not, you may want to consider removing it.

It may be mandatory for some use case as defined in other documents.  The card driver is verifying the CVC(s)  which are stored on the card in hard to access objects. The "Secure Messaging Certificate Signer" certificate is easily  accessed  by  calling application via pkcs11 or pkcs15.  For example:
```
/pkcs15-tool --read-certificate 25 
Using reader with a card: Identiv SCR3500 A Contact Reader [CCID Interface] (54302140601312) 00 00
Connecting to card in reader Identiv SCR3500 A Contact Reader [CCID Interface] (54302140601312) 00 00...
Using card driver Personal Identity Verification Card.
-----BEGIN CERTIFICATE-----
MIIFCTCCA/GgAwIBAgIBLDANBgkqhkiG9w0BAQsFADCBtDELMAkGA1UEBhMCRlIx
HjAcBgNVBAoMFU9CRVJUSFVSIFRFQ0hOT0xPR0lFUzEcMBoGA1UECwwTMDAwMiAz
NDA3MDk1MzQwMDI1OTEMMAoGA1UECwwDQ0FJMRUwEwYDVQQLDAxSbkQgSURFTlRJ
VFkxQjBABgNVBAMMOVRFU1QgUlNBIDIwNDggQ0EgRk9SIE9CRVJUSFVSIFRFQ0hO
T0xPR0lFUyBURVNUIFBJViBDQVJEUzAeFw0xNzA1MTYwNzExNDhaFw0zNzA1MTYw
NzExNDhaMIGsMQswCQYDVQQGEwJGUjEeMBwGA1UECgwVT0JFUlRIVVIgVEVDSE5P
TE9HSUVTMRwwGgYDVQQLDBMwMDAyIDM0MDcwOTUzNDAwMjU5MQwwCgYDVQQLDAND
QUkxFTATBgNVBAsMDFJuRCBJREVOVElUWTE6MDgGA1UEAwwxVEVTVCBQSVYgQ09O
VEVOVCBTSUdORVIgUlNBIEZPUiBPVCBURVNUIFBJViBDQVJEUzCCASIwDQYJKoZI
hvcNAQEBBQADggEPADCCAQoCggEBAMr3EWrES919KE0sy+Yii+/JGxc6Dxqveczn
zeJ65WhlTNnBkpu3f7iYxOYbIOA29CUWh8DLKcuJWhbEIqQeEoH0tTufGJmqKjbO
eewGfRLVEtqwdxBX12eRvYfoz58ALGTHXUKAO9dxQkc7XsbEPl8iL90zhg+L1Jvo
qF301hCkI5FhQ6z7KYmGxMUUyFEkni5RhMxw8gBmeB9pV9N0X61FaZvcj1EAfuuj
leEjd6eq0OrLnXM2HGmGukn1Rj2es5oQ1HSank6HzMs+jDdAtIz99ycbsa7zGV8D
NzA8joUj07PMJEFvPKzoVTQzwlrzqfMtRogalm4dsZo3e9g+cKMCAwEAAaOCASow
ggEmMEcGA1UdHwRAMD4wPKA6oDiGNmh0dHBzOi8vZGVtb3Bpdi5henVyZXdlYnNp
dGVzLm5ldC9kZW1vcGl2Y2Fyc2EyMDQ4LmNybDBSBggrBgEFBQcBAQRGMEQwQgYI
KwYBBQUHMAKGNmh0dHBzOi8vZGVtb3Bpdi5henVyZXdlYnNpdGVzLm5ldC9kZW1v
cGl2Y2Fyc2EyMDQ4LnA3YzAdBgNVHQ4EFgQURwC+ngAbbQ7GDP1lx2Br+Yffn8Mw
HwYDVR0jBBgwFoAUliWr0whTaY+fsL0AFdrgz91GFJkwCQYDVR0TBAIwADAOBgNV
HQ8BAf8EBAMCBsAwFwYDVR0gBBAwDjAMBgpghkgBZQMCAQMnMBMGA1UdJQQMMAoG
CGCGSAFlAwYHMA0GCSqGSIb3DQEBCwUAA4IBAQBiWz1AUA3MoMj76wosaXtN4rZI
P8qUXWoJw+RDWXGmiatsO1peOjeTeJAUAlOWJ19UJaCfn84vuYjTo6Xci/nzlAQ9
yifr26izLLeIpCSPcbHJVnRyudkeAjthyVmmVEuh9GX5OyVUqL6RKQ5m3q9KNlOg
uByqyrqx0Y+o1a564fUyncc8SPHruYuS3hwoWYGLpPFZssXDZcz1ncqLf316ZZUd
3LboMZxMVhD4prtYDK9OEO3BHJNHlyw2Q4I3f/bRuC3ELo89MtavH/2FhFXEQfWJ
3W5pDDtlcrktERz3jKk+Uu6iPBTzvpjOCo5DSdaOBAZrP1fXimKGd0MRs0Y2
-----END CERTIFICATE-----
```

I would prefer to leave that code in as the CVCs are only verified during the  "4.1 The Key Establishment Protocol"
which in most cases is only done once per insertion of the card. 

Leaving the code alone allows the calling application to do any additional verification including CRL or OSCP if required by some policies. 

We really don't know: "c) implies managing a trust anchor by the user to make it actually meaningful (which nobody is interested in)."  PIV was designed to be administered by government or large organizations that are interested in   managing trust anchors **for** their users.


The comments in  lines 2064 - 2080 where asking similar questions

Reply:
**This implementation** retrieves an X.509 certificate from the card and checks checks it against a CVC that was retrieved from the card. You don't have any independent anchor of trust. If some rogue card comes with bad CVCs, it can supply a bad X.509 trust anchor as well. Checking the signature is nothing more that an easily forged sanity check. From a security perspective, this check doesn't make sense.

As you know, I like simple and clean implementations. But as said before, I'll leave the decision to you.

Reply:
> As you know, I like simple and clean implementations. But as said before, I'll leave the decision to you.

Leaving in for now.